File: day5/python/day5.py
import os

os.chdir("..")
data = open(os.getcwd()+"/input.txt", "r").read().split("\n\n")


def crate() -> list[list[str]]:
    box = []
    box.append([i for i in "FLMW"[::-1]])
    box.append([i for i in "FMVZB"[::-1]])
    box.append([i for i in "QLSRVH"[::-1]])
    box.append([i for i in "JTMPQVSF"[::-1]])
    box.append([i for i in "WSL"[::-1]])
    box.append([i for i in "WJRMPVF"[::-1]])
    box.append([i for i in "FRNPCQJ"[::-1]])
    box.append([i for i in "BRWZSPHV"[::-1]])
    box.append([i for i in "WZHGCJMB"[::-1]])
    return box


def instruction(data: str):
    return [[int(j) for j in i.split(" ") if j.isdigit()] for i in data.split("\n")]


def main():
    inst = instruction(data[1])
    crt = crate()
    for i in inst:
        for j in range(i[0]):
            crt[i[2]-1].append(crt[i[1]-1].pop())
    print("stage 1 = ", "".join([i[-1] for i in crt]))
    crt2 = crate()
    for i in inst:
        crt2[i[2]-1].extend(crt2[i[1]-1][-(i[0]):])
        crt2[i[1]-1] = crt2[i[1]-1][:-(i[0])]
    print("stage 2 = ", "".join([i[-1] for i in crt2]))


print(main())

File: day5/golang/day5.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	stage1()
	stage2()
}

func stage1() {
	inst := instruction()
	crate := inputCrate()
	for _, i := range inst {
		for j := 0; j < i[0]; j++ {
			length := len(crate[i[1]-1])
			crate[i[2]-1] = append(crate[i[2]-1], crate[i[1]-1][length-1])
			crate[i[1]-1] = crate[i[1]-1][:length-1]
		}
	}
	fmt.Println("stage 1 =", lastAlp(crate))
}
func stage2() {
	inst := instruction()
	crate := inputCrate()
	for _, i := range inst {
		length := len(crate[i[1]-1])
		for j := i[0]; j > 0; j-- {
			crate[i[2]-1] = append(crate[i[2]-1], crate[i[1]-1][length-j])
		}
		crate[i[1]-1] = crate[i[1]-1][:length-i[0]]
	}
	fmt.Println("stage 2 =", lastAlp(crate))
}
func lastAlp(in [][]rune) string {
	var out []rune
	for _, i := range in {
		out = append(out, i[len(i)-1])
	}
	return string(out)
}

func inputCrate() [][]rune {
	var input [][]rune
	input = append(input, reverse("FLMW"))
	input = append(input, reverse("FMVZB"))
	input = append(input, reverse("QLSRVH"))
	input = append(input, reverse("JTMPQVSF"))
	input = append(input, reverse("WSL"))
	input = append(input, reverse("WJRMPVF"))
	input = append(input, reverse("FRNPCQJ"))
	input = append(input, reverse("BRWZSPHV"))
	input = append(input, reverse("WZHGCJMB"))
	return input
}
func instruction() [][]int {
	var input [][]int
	file, _ := os.ReadFile("../input.txt")
	inst := strings.Split(string(file), "\n\n")
	for _, i := range strings.Split(inst[1], "\n") {
		var hold []int
		for _, j := range strings.Split(i, " ") {
			integer, err := strconv.Atoi(j)
			if err == nil {
				hold = append(hold, integer)
			}
		}
		input = append(input, hold)
	}
	return input
}
func reverse(in string) []rune {
	var result string
	for _, i := range in {
		result = string(i) + result
	}
	return []rune(result)
}
